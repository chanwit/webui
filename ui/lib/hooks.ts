import _ from "lodash";
import qs from "query-string";
import { useContext, useEffect, useState } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { AppContext } from "../components/AppStateProvider";
import {
  Context,
  HelmRelease,
  Kustomization,
  Source,
  Workload,
} from "./rpc/clusters";
import { AllNamespacesOption } from "./types";
import { clustersClient, formatURL, normalizePath, PageRoute } from "./util";
// The backend doesn't like the word "all". Instead, it wants an empty string.
// Navigation might get weird if we use an empty string on the front-end.
// There may also be a naming collision with a namespace named "all".
const formatAPINamespace = (ns: string) =>
  ns === AllNamespacesOption ? "" : ns;

export function useKubernetesContexts(): {
  contexts: Context[];
  namespaces: string[];
  currentContext: string;
  currentNamespace: string;
} {
  const { navigate } = useNavigation();
  const { context, namespace } = qs.parse(location.search) as {
    context: string;
    namespace: string;
  };
  const { contexts, namespaces } = useContext(AppContext);

  useEffect(() => {
    if (!context) {
      navigate(PageRoute.Redirector, null, null);
    }
  }, []);

  return {
    contexts,
    namespaces: namespaces[context] || [],
    currentContext: context,
    currentNamespace: namespace,
  };
}

type KustomizationList = { [name: string]: Kustomization };

export function useKustomizations(
  currentContext: string,
  currentNamespace: string
) {
  const { doError } = useContext(AppContext);
  const [kustomizations, setKustomizations] = useState({} as KustomizationList);

  useEffect(() => {
    if (!currentContext) {
      return;
    }

    clustersClient
      .listKustomizations({
        contextname: currentContext,
        namespace: formatAPINamespace(currentNamespace),
      })
      .then((res) => {
        const r = _.keyBy(res.kustomizations, "name");
        setKustomizations(r);
      })
      .catch((err) => {
        doError(
          "There was an error fetching kustomizations",
          true,
          err.message
        );
      });
  }, [currentContext, currentNamespace]);

  const syncKustomization = (k: Kustomization) =>
    clustersClient
      .syncKustomization({
        contextname: currentContext,
        namespace: k.namespace,
        withsource: false,
        kustomizationname: k.name,
      })
      .then(() => {
        setKustomizations({
          ...kustomizations,
          [k.name]: k,
        });
      });

  return { kustomizations, syncKustomization };
}

export enum SourceType {
  Git = "git",
  Bucket = "bucket",
  Helm = "helm",
  Chart = "chart",
}

function convertSourceTypeToInt(s: SourceType) {
  switch (s) {
    case SourceType.Git:
      return 0;
    case SourceType.Bucket:
      return 1;
    case SourceType.Helm:
      return 2;
    case SourceType.Chart:
      return 3;
  }
}

type SourceData = {
  [SourceType.Git]: Source[];
  [SourceType.Bucket]: Source[];
  [SourceType.Helm]: Source[];
};

export function useSources(
  currentContext: string,
  currentNamespace: string
): SourceData {
  const { doError } = useContext(AppContext);
  const [sources, setSources] = useState({
    [SourceType.Git]: [],
    [SourceType.Bucket]: [],
    [SourceType.Helm]: [],
  });

  useEffect(() => {
    if (!currentContext) {
      return;
    }

    const p = _.map(SourceType, (s) =>
      clustersClient.listSources({
        contextname: currentContext,
        namespace: formatAPINamespace(currentNamespace),
        // @ts-ignore
        sourcetype: convertSourceTypeToInt(s),
      })
    );

    Promise.all(p)
      .then((arr) => {
        const d = {};
        _.each(arr, (a) => {
          _.each(a.sources, (src) => {
            const t = _.lowerCase(src.type);
            if (!d[t]) {
              d[t] = [];
            }

            d[t].push(src);
          });
        });
        setSources(d as SourceData);
      })
      .catch((err) => {
        doError("There was an error fetching sources", true, err);
      });
  }, [currentContext, currentNamespace]);

  return sources;
}

export function useHelmReleases(
  currentContext: string,
  currentNamespace: string
): { [name: string]: HelmRelease } {
  const [helmReleases, setHelmReleases] = useState({});
  const { doError } = useContext(AppContext);

  useEffect(() => {
    if (!currentContext) {
      return;
    }

    clustersClient
      .listHelmReleases({
        contextname: currentContext,
        namespace: formatAPINamespace(currentNamespace),
      })
      .then((res) => {
        const releases = _.keyBy(res.helmReleases, "name");
        setHelmReleases(releases);
      })
      .catch((err) => {
        doError("There was an error fetching helm releases", true, err.message);
      });
  }, [currentContext, currentNamespace]);

  return helmReleases;
}

export function useAppState() {
  const { appState } = useContext(AppContext);
  return appState;
}

export function useNavigation() {
  const history = useHistory();
  const location = useLocation();
  const [currentPage, setCurrentPage] = useState("");

  useEffect(() => {
    const [pageName] = normalizePath(location.pathname);
    setCurrentPage(pageName as string);
  }, [location.pathname]);

  return {
    currentPage,
    query: qs.parse(location.search),
    navigate: (
      page: PageRoute | null,
      context: string,
      namespace: string,
      query: any = {}
    ) => {
      let nextPage = page || currentPage;

      if (nextPage == "error") {
        nextPage = PageRoute.Home;
      }

      history.push(formatURL(nextPage as string, context, namespace, query));
    },
  };
}

export function useWorkloads(currentContext: string, currentNamespace: string) {
  const [workloads, setWorkloads] = useState([] as Workload[]);

  useEffect(() => {
    if (!currentContext || !currentNamespace) {
      return;
    }

    clustersClient
      .listWorkloads({
        contextname: currentContext,
        namespace:
          currentNamespace === AllNamespacesOption ? "" : currentNamespace,
      })
      .then((res) => {
        setWorkloads(res.workloads);
      });
  }, [currentContext, currentNamespace]);
  return workloads;
}
