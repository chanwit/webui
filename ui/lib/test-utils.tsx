import { MuiThemeProvider } from "@material-ui/core";
import _ from "lodash";
import * as React from "react";
import { MemoryRouter } from "react-router";
import { ThemeProvider } from "styled-components";
import AppStateProvider from "../components/AppStateProvider";
import { ListContextsRes, ListNamespacesForContextRes } from "./rpc/clusters";
import theme from "./theme";

type ClientOverrides = {
  listContexts: ListContextsRes;
  listNamespacesForContext: ListNamespacesForContextRes;
};

export function withContext(
  TestComponent,
  simulatedUrl: string,
  clientOverrides: ClientOverrides
) {
  // Don't make the user wire up all the promise stuff to be interface-compliant
  const promisified = _.reduce(
    clientOverrides,
    (result, desiredResponse, method) => {
      result[method] = () =>
        new Promise((accept) => accept(desiredResponse as any));

      return result;
    },
    {}
  );

  return (
    <MuiThemeProvider theme={theme}>
      <ThemeProvider theme={theme}>
        <MemoryRouter initialEntries={[simulatedUrl]}>
          <AppStateProvider clustersClient={promisified}>
            <TestComponent />
          </AppStateProvider>
        </MemoryRouter>
      </ThemeProvider>
    </MuiThemeProvider>
  );
}
