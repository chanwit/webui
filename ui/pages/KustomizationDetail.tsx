import {
  Box,
  Button,
  CircularProgress,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@material-ui/core";
import _ from "lodash";
import * as React from "react";
import styled from "styled-components";
import Flex from "../components/Flex";
import KeyValueTable from "../components/KeyValueTable";
import Link from "../components/Link";
import Panel from "../components/Panel";
import {
  useKubernetesContexts,
  useKustomizations,
  useNavigation,
  useWorkloads,
} from "../lib/hooks";
import { Kustomization } from "../lib/rpc/clusters";
import { formatURL, PageRoute } from "../lib/util";

type Props = {
  className?: string;
};

const Styled = (c) => styled(c)``;

const infoFields = [
  "sourceref",
  "namespace",
  "reconcileat",
  "path",
  "interval",
  "prune",
  "reconcilerequestat",
];

const formatInfo = (detail: Kustomization) =>
  _.map(_.pick(detail, infoFields), (v, k) => ({
    key: k,
    value: typeof v === "string" ? v : v.toString(),
  }));

function KustomizationDetail({ className }: Props) {
  const [syncing, setSyncing] = React.useState(false);
  const { query } = useNavigation();
  const { currentContext, currentNamespace } = useKubernetesContexts();
  const workloads = useWorkloads(currentContext, currentNamespace);

  const { kustomizations, syncKustomization } = useKustomizations(
    currentContext,
    currentNamespace
  );
  const kustomizationDetail = kustomizations[query.kustomizationId as string];

  const handleSyncClicked = () => {
    setSyncing(true);

    syncKustomization(kustomizationDetail).then(() => {
      setSyncing(false);
    });
  };

  if (!kustomizationDetail) {
    return null;
  }

  const overrides = {
    sourceref: [
      <Link
        to={formatURL(
          PageRoute.SourceDetail,
          currentContext,
          currentNamespace,
          {
            sourceType: kustomizationDetail.sourcerefkind.toLowerCase(),
            sourceId: kustomizationDetail.sourceref,
          }
        )}
      >
        {kustomizationDetail.sourceref}
      </Link>,
      "Source",
    ],
    reconcileat: [
      ` ${new Date(
        kustomizationDetail.reconcileat
      ).toLocaleTimeString()} ${new Date(
        kustomizationDetail.reconcileat
      ).toLocaleDateString()}`,
      "Last Reconcile",
    ],
    reconcilerequestat: [
      ` ${new Date(
        kustomizationDetail.reconcilerequestat
      ).toLocaleTimeString()} ${new Date(
        kustomizationDetail.reconcilerequestat
      ).toLocaleDateString()}`,
      "Last Reconcile Request",
    ],
  };

  return (
    <div className={className}>
      <Box m={2}>
        <Flex align center wide>
          <Flex wide>
            <h2>{kustomizationDetail.name}</h2>
          </Flex>
          <Button
            onClick={handleSyncClicked}
            color="primary"
            disabled={syncing}
            variant="contained"
          >
            {syncing ? <CircularProgress size={24} /> : "Sync"}
          </Button>
        </Flex>
        <Panel title="Info">
          <KeyValueTable
            columns={4}
            pairs={formatInfo(kustomizationDetail)}
            overrides={overrides}
          />
        </Panel>
      </Box>

      <Box m={2}>
        <Panel title="Conditions">
          <TableContainer>
            <Table>
              <TableHead>
                <TableRow>
                  <TableCell>Type</TableCell>
                  <TableCell>Status</TableCell>
                  <TableCell>Timestamp</TableCell>
                  <TableCell>Reason</TableCell>
                  <TableCell>Message</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {_.map(kustomizationDetail.conditions, (c, i) => (
                  <TableRow key={i}>
                    <TableCell>{c.type}</TableCell>
                    <TableCell>{c.status}</TableCell>
                    <TableCell>
                      {new Date(c.timestamp).toLocaleDateString()}
                    </TableCell>
                    <TableCell>{c.reason}</TableCell>
                    <TableCell>{c.message}</TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </Panel>
      </Box>
      <Box m={2}>
        <Panel title="Related Workloads">
          {_.map(workloads, (w) => {
            return (
              <div key={w.name}>
                <Link
                  to={formatURL(
                    PageRoute.WorkloadDetail,
                    currentContext,
                    w.namespace,
                    { workloadId: w.name }
                  )}
                >
                  {w.name}
                </Link>
              </div>
            );
          })}
        </Panel>
      </Box>
    </div>
  );
}

export default Styled(KustomizationDetail);
