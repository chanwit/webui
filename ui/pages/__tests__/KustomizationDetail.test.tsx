import { render, screen } from "@testing-library/react";
import { stubbedClustersResponses, withContext } from "../../lib/test-utils";
import KustomizationDetail from "../KustomizationDetail";

describe("KustomizationDetail", () => {
  let container;
  beforeEach(() => {
    container = document.createElement("div");
    document.body.appendChild(container);
  });
  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });
  it("renders", async () => {
    const url =
      "/?context=my-context&namespace=default&kustomizationId=my-kustomization";
    const tree = render(
      withContext(KustomizationDetail, url, stubbedClustersResponses),
      container
    );

    expect(
      (await screen.findByText("my-kustomization")).textContent
    ).toBeTruthy();
    expect(tree).toMatchSnapshot();
  });
});
