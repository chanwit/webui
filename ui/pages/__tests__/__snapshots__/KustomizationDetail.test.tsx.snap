// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KustomizationDetail renders 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="sc-jSgupP sc-fubCfw ERjJi"
      >
        <div
          class="sc-bdfBwQ kLrCXV"
        >
          <nav
            class="MuiTypography-root MuiBreadcrumbs-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            <ol
              class="MuiBreadcrumbs-ol"
            >
              <li
                class="MuiBreadcrumbs-li"
              >
                <a
                  class="sc-eCssSg"
                  href="/kustomizations?context=my-context&namespace=all"
                >
                  <h2>
                    Kustomizations
                  </h2>
                </a>
              </li>
              <li
                aria-hidden="true"
                class="MuiBreadcrumbs-separator"
              >
                /
              </li>
              <li
                class="MuiBreadcrumbs-li"
              >
                <div
                  class="sc-bdfBwQ jANNfu"
                >
                  <h2>
                    my-kustomization
                  </h2>
                </div>
              </li>
            </ol>
          </nav>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Sync
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root MuiBox-root-1"
        >
          <div
            class="sc-iBPRYJ"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
            >
              <div
                class="sc-gKsewC ejJyga"
              >
                <h3>
                  Info
                </h3>
              </div>
              <div
                class="MuiCardContent-root"
              >
                <div>
                  <div
                    class="sc-hKgILt emSDXW"
                    role="list"
                  >
                    <table>
                      <tbody>
                        <tr>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Source"
                              class="sc-gsTCUz bCQwq"
                            >
                              Source
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                              <a
                                class="sc-eCssSg"
                                href="/sources_detail?context=my-context&namespace=all&sourceId=my-source&sourceType=git"
                              >
                                my-source
                              </a>
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Namespace"
                              class="sc-gsTCUz bCQwq"
                            >
                              Namespace
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                              default
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Last reconcile"
                              class="sc-gsTCUz bCQwq"
                            >
                              Last reconcile
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                               Invalid Date Invalid Date
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Path"
                              class="sc-gsTCUz bCQwq"
                            >
                              Path
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                              /k8s
                            </div>
                          </td>
                        </tr>
                        <tr>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Interval"
                              class="sc-gsTCUz bCQwq"
                            >
                              Interval
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                              1m
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Prune"
                              class="sc-gsTCUz bCQwq"
                            >
                              Prune
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                              false
                            </div>
                          </td>
                          <td
                            role="listitem"
                          >
                            <div
                              aria-label="Last reconcile request"
                              class="sc-gsTCUz bCQwq"
                            >
                              Last reconcile request
                            </div>
                            <div
                              class="sc-dlfnbm"
                            >
                               Invalid Date Invalid Date
                            </div>
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-2"
        >
          <div
            class="sc-iBPRYJ"
          >
            <div
              class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
            >
              <div
                class="sc-gKsewC ejJyga"
              >
                <h3>
                  Conditions
                </h3>
              </div>
              <div
                class="MuiCardContent-root"
              >
                <div>
                  <div
                    class="MuiTableContainer-root"
                  >
                    <table
                      class="MuiTable-root"
                    >
                      <thead
                        class="MuiTableHead-root"
                      >
                        <tr
                          class="MuiTableRow-root MuiTableRow-head"
                        >
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Type
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Status
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Timestamp
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Reason
                          </th>
                          <th
                            class="MuiTableCell-root MuiTableCell-head"
                            scope="col"
                          >
                            Message
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="MuiTableBody-root"
                      />
                    </table>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jSgupP sc-fubCfw ERjJi"
    >
      <div
        class="sc-bdfBwQ kLrCXV"
      >
        <nav
          class="MuiTypography-root MuiBreadcrumbs-root MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          <ol
            class="MuiBreadcrumbs-ol"
          >
            <li
              class="MuiBreadcrumbs-li"
            >
              <a
                class="sc-eCssSg"
                href="/kustomizations?context=my-context&namespace=all"
              >
                <h2>
                  Kustomizations
                </h2>
              </a>
            </li>
            <li
              aria-hidden="true"
              class="MuiBreadcrumbs-separator"
            >
              /
            </li>
            <li
              class="MuiBreadcrumbs-li"
            >
              <div
                class="sc-bdfBwQ jANNfu"
              >
                <h2>
                  my-kustomization
                </h2>
              </div>
            </li>
          </ol>
        </nav>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Sync
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-1"
      >
        <div
          class="sc-iBPRYJ"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="sc-gKsewC ejJyga"
            >
              <h3>
                Info
              </h3>
            </div>
            <div
              class="MuiCardContent-root"
            >
              <div>
                <div
                  class="sc-hKgILt emSDXW"
                  role="list"
                >
                  <table>
                    <tbody>
                      <tr>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Source"
                            class="sc-gsTCUz bCQwq"
                          >
                            Source
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                            <a
                              class="sc-eCssSg"
                              href="/sources_detail?context=my-context&namespace=all&sourceId=my-source&sourceType=git"
                            >
                              my-source
                            </a>
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Namespace"
                            class="sc-gsTCUz bCQwq"
                          >
                            Namespace
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                            default
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Last reconcile"
                            class="sc-gsTCUz bCQwq"
                          >
                            Last reconcile
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                             Invalid Date Invalid Date
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Path"
                            class="sc-gsTCUz bCQwq"
                          >
                            Path
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                            /k8s
                          </div>
                        </td>
                      </tr>
                      <tr>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Interval"
                            class="sc-gsTCUz bCQwq"
                          >
                            Interval
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                            1m
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Prune"
                            class="sc-gsTCUz bCQwq"
                          >
                            Prune
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                            false
                          </div>
                        </td>
                        <td
                          role="listitem"
                        >
                          <div
                            aria-label="Last reconcile request"
                            class="sc-gsTCUz bCQwq"
                          >
                            Last reconcile request
                          </div>
                          <div
                            class="sc-dlfnbm"
                          >
                             Invalid Date Invalid Date
                          </div>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root MuiBox-root-2"
      >
        <div
          class="sc-iBPRYJ"
        >
          <div
            class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
          >
            <div
              class="sc-gKsewC ejJyga"
            >
              <h3>
                Conditions
              </h3>
            </div>
            <div
              class="MuiCardContent-root"
            >
              <div>
                <div
                  class="MuiTableContainer-root"
                >
                  <table
                    class="MuiTable-root"
                  >
                    <thead
                      class="MuiTableHead-root"
                    >
                      <tr
                        class="MuiTableRow-root MuiTableRow-head"
                      >
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Type
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Status
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Timestamp
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Reason
                        </th>
                        <th
                          class="MuiTableCell-root MuiTableCell-head"
                          scope="col"
                        >
                          Message
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="MuiTableBody-root"
                    />
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
